src/Swarm/Language/Typecheck.hs

```diff
@@ -453,8 +453,10 @@ infer s@(Syntax l t) = (`catchError` addLocToTypeErr s) $ case t of
     _ <- decomposeCmdTy (c2' ^. sType)
     return $ Syntax' l (SBind mx c1' c2') (c2' ^. sType)
   SAnnotate c pty -> do
-    uty <- skolemize $ toU pty
-    check c uty
+    uty <- instantiate $ toU pty
```

Comment:
Right, this isn't going to work because it allows things like `3 : a`, as you noted.  If `a` turns into a unification variable then it will be free to unify with `int`.  We really do need `skolemize` here for checking the type of `c`, so that we can check that it really is polymorphic as advertised.  But then we need to throw away the type we got from checking `c` and return a version which has been `instantiate`d.  So maybe something like this (untested):

```
sty <- skolemize $ toU pty
uty <- instantiate $ toU pty
c' <- check c sty
return $ Syntax' l (SAnnotate c' pty) uty
```

An alternative would be to just say that polytypes are not allowed in type annotations.  That would make things a lot simpler.  I don't have a good sense of how big/annoying of a restriction that would be.

Reply:
Thank you for this explanation. In fact I have just pushed a commit that is very similar to your proposed solution (which I have just seen afterwards). What really helped me understand what is going on here is that I read through your rather detailed and (at least in my opinion) well-written [blog post](https://byorgey.wordpress.com/2021/09/08/implementing-hindley-milner-with-the-unification-fd-library) on your Hendley-Milner implementation. 

I have added a few more test cases to validate the correctness of my last change. Essentially it only differes from your proposal by the presence of a check for escaping Skolem variables. Without it, `\f. (f:forall a. a->a) 3` would typecheck, which it really shouldn't without higher ranked types. If I understand it correctly, this would be the same problem as in #211.

Reply:
I see another difference now which puzzles me a little bit. You only check `c` against `sty` and use the result in the resulting AST node. I did something similar in an intermediate version of my implementation, but wondered whether this is a good idea, since `check` puts the unified type into the resulting AST node. So wouldn't `Syntax' l (SAnnotate c' pty) uty` lead to a possibly problematic missmatch between `uty` and `c' ^. sType` ?

Reply:
Ah, glad you found the blog post useful!

Checking for escaping skolems makes sense, and of course you're right about the potential problems with my version having a mismatch between `uty` and `c' ^. sType`.  I think the way you have done it makes sense.  In an abstract sense it bothers me somewhat to call `check` twice on `c`, once with a skolemized type and once with an instantiated type---but I don't actually know of a better way to do it!